jerry-core/ecma/operations/ecma-atomics-object.c

```diff
@@ -0,0 +1,253 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ecma-atomics-object.h"
+#include "ecma-arraybuffer-object.h"
+#include "ecma-shared-arraybuffer-object.h"
+#include "ecma-bigint.h"
+#include "ecma-typedarray-object.h"
+#include "ecma-objects.h"
+#include "ecma-builtins.h"
+#include "ecma-exceptions.h"
+#include "ecma-gc.h"
+#include "ecma-globals.h"
+#include "ecma-helpers.h"
+#include "jmem.h"
+#include "jcontext.h"
+#include "ecma-function-object.h"
+
+#if JERRY_BUILTIN_ATOMICS
+
+/** \addtogroup ecma ECMA
+ * @{
+ *
+ * \addtogroup ecmaatomicsobject ECMA builtin Atomics helper functions
+ * @{
+ */
+
+/**
+ * Atomics validate Shared integer typedArray
+ *
+ * See also: ES11 24.4.1.1
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_validate_shared_integer_typedarray (ecma_value_t typedarray,
+                                         bool waitable)
+{
+  /* 2. */
+  if (!ecma_is_value_object (typedarray))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not an object"));
+  }
+
+  /* 3-4. */
+  ecma_object_t *typedarray_p = ecma_get_object_from_value (typedarray);
+  ecma_typedarray_info_t target_info =  ecma_typedarray_get_info (typedarray_p);
+
+  /* 5-6. */
+  if (waitable)
+  {
+    if (!(target_info.id == ECMA_BIGINT64_ARRAY || target_info.id == ECMA_INT32_ARRAY))
+    {
+      return ecma_raise_type_error (ECMA_ERR_MSG ("Argument is not supported"));
+    }
+  }
+  else
+  {
+    if (target_info.id  == ECMA_UINT8_CLAMPED_ARRAY || target_info.id == ECMA_FLOAT32_ARRAY
+        || target_info.id == ECMA_FLOAT64_ARRAY)
+    {
+      return ecma_raise_type_error (ECMA_ERR_MSG ("Argument is not supported"));
+    }
+  }
+
+  /* 7. */
+  JERRY_ASSERT (target_info.array_buffer_p != NULL);
+
+  /* 8-10. */
+  ecma_object_t *buffer = ecma_typedarray_get_arraybuffer (typedarray_p);
+
+  if (!ecma_object_class_is (buffer,ECMA_OBJECT_CLASS_SHARED_ARRAY_BUFFER))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument is not SharedArrayBuffer"));
+  }
+
+  return ecma_make_object_value (buffer);
+} /* ecma_validate_shared_integer_typedarray */
+
+/**
+ * Atomics validate Atomic Access
+ *
+ * See also: ES11 24.4.1.2
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_validate_atomic_access (ecma_value_t typedarray,
+                             ecma_value_t requestIndex)
+{
+  /* 1. */
+  JERRY_ASSERT (ecma_is_value_object (typedarray)
+                && ecma_typedarray_get_arraybuffer (ecma_get_object_from_value (typedarray)) != NULL);
+
+  ecma_object_t *typedarray_p = ecma_get_object_from_value (typedarray);
+
+  /* 2. */
+  ecma_number_t access_index;
+  ecma_op_to_index (requestIndex, &access_index);
```

Comment:
* `ecma_op_to_index` has a return value which should not be ignored.
* There should be a test case where the op_to_index fails.